ledger/complete/mtrie/trie/trie.go

```diff
@@ -379,70 +385,82 @@ type updateResult struct {
 //     (excluding the bit at index headHeight)
 //   - paths are NOT duplicated
 func update(
-	nodeHeight int, parentNode *node.Node,
-	paths []ledger.Path, payloads []ledger.Payload, compactLeaf *node.Node,
-	prune bool,
+	nodeHeight int, // the height of the node during traversing the subtree
+	currentNode *node.Node, // the current node on the travesing path, if it's nil it means the trie has no node on this path
+	paths []ledger.Path, // the paths to update the payloads
```

Comment:
looks like a lot of cycle is wasted over iterating over paths/payloads.  the leaf node update is particularly messy.  Consider replacing the paths, payloads arguments with map[Path]Payload

Reply:
or list of sorted dedup pair

Reply:
>looks like a lot of cycle is wasted over iterating over paths/payloads. 

I don't understand the wasted iteration you are talking about. Can you explain a bit more?

Note, when traversing the tree, we need to split and carry different paths and payloads to continue traverse along the subtrees on different branches. So splitByPath does the split, as well as in-place sort. I'm not sure how using a map would save any iteration cycle. Wouldn't a map need to be split and carried along different branches?

Reply:
we're doing a linear scan at line 427 to match the path.  we can replace this either with a hash lookup (if input is dict), or a bisection search (if input is sorted pairs).

Reply:
Good point 👍